perfmetrics/scripts/load_tests/python/tasks/tf_data.py

```diff
@@ -0,0 +1,138 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Contains tasks which reads data from files using tensorflow's tf.data api.
+
+Example:
+
+  lg_obj = load_generator.LoadGenerator(...)
+  task_obj = tf_data.TFDataRead(...)
+  observations = lg_obj.generate_load(task_obj)
+"""
+import os
+import logging
+import tensorflow as tf
+from load_generator import task
+
+# .tfrecord file contains multiple TFRecords in it. This defines the size of
+# that single TFRecord in bytes.
+SINGLE_RECORD_SIZE = 256 * 1024
+TF_DATA_READ = 'tf_data_read'
+
+
+class TFDataRead(task.LoadTestTask):
+  """Task class for reading tfrecord file using tensorflow's tf.data api.
+
+  Note: The same class can be used with tf's internal GCS client if format of
+  file path starts with gs:// and with GCSFuse if local GCSFuse mounted path
+  is passed.
+
+  For details on prefetch, num_parallel_calls & shard arguments, please refer
+  to tf.data public documentation.
+
+  Args:
+    task_name: String name assigned to task.
+    file_path_format: String format of file path. Must contain 'file_num' to be
+      formatted. E.g. gcs/256kb/read.{file_num}.
+    file_size: Integer size of file in bytes to be read.
+    num_files: Integer number of files to be read in one task.
+    prefetch: Integer value of prefetch. Default to AUTOTUNE(-1).
+    num_parallel_calls = Integer value of num_parallel_calls. Default to
+      AUTOTUNE(-1).
+    shard: Integer value of shard. Default to AUTOTUNE(-1).
+  """
+
+  def __init__(self,
+               task_name,
+               file_path_format,
+               file_size,
+               num_files,
+               prefetch=-1,
+               num_parallel_calls=-1,
+               shard=1):
+    super().__init__()
+    self.task_type = TF_DATA_READ
+    self.task_name = task_name
+    self.file_path_format = file_path_format
+    self.file_size = file_size
+    self.num_files = num_files
+    self.prefetch = prefetch
+    self.num_parallel_calls = num_parallel_calls
+    self.shard = shard
+    self.file_names = [
+        self.file_path_format.format(file_num=file_num)
+        for file_num in range(self.num_files)
+    ]
+
+  def pre_task(self, process_id, thread_id):  #pylint: disable=unused-argument
+    """Pre-task to clear OS's page caches in load testing.
+    """
+    # Clear the page cache as there is no way to bypass cache in tf.data.
+    # Note: This takes time and hence decrease the average bandwidth.
+    os.system("sudo sh -c 'sync; echo 3 >  /proc/sys/vm/drop_caches'")
+
+  def task(self, process_id, thread_id):  #pylint: disable=unused-argument
+    """Reads TFRecord files using tensorflow's tf.data method.
+
+    See base class for more details.
+
+    Returns:
+      Integer denoting the size of content read in bytes.
+    """
+    content_len = 0
+    files_dataset = tf.data.Dataset.from_tensor_slices(self.file_names)
+
+    def tfrecord(path):
```

Comment:
Can we add the documentation we referred while creating shard and all the concepts related to shard?

Reply:
Added in internal docs and added link to tf.data.TFRecordDataset.shard